### PR TITLE
[9.0][FIX] account_voucher. Link between payment and invoice

### PR DESCRIPTION
**Pull request by AaronHForgeFlow**
Solves this issue: https://github.com/OCA/OpenUpgrade/issues/1283

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Next time, please publish directly the PR with the issue text inside, instead of making 2 things that are worst to be followed. About this PR, please check Travis.

**Comment by AaronHForgeFlow** [comment]
This is working fine for me, but creating an extra column maybe is not the best way to do it. @NL66278 @jbeficent

File: addons/account_voucher/migrations/9.0.1.0/post-migration.py [hunk 1]
```diff
@@ -83,15 +83,56 @@ def create_payments_from_vouchers(env):
         AND aml.id = aml2.id
         """
     )
+    env.cr.execute("""SELECT column_name
```
Comment (pedrobaeza) [inline addons/account_voucher/migrations/9.0.1.0/post-migration.py]: Use `openupgradelib.column_exists`

File: addons/account_voucher/migrations/9.0.1.0/post-migration.py [hunk 2]
```diff
@@ -83,15 +83,56 @@ def create_payments_from_vouchers(env):
...
+        column_name='payment_from_invoice_id'""")
+    if not env.cr.fetchone():
+        env.cr.execute("""
+            ALTER TABLE account_move_line ADD COLUMN
+            payment_from_invoice_id INTEGER
+            """)
+
+    env.cr.execute(
+        """\
+        WITH Q1 AS (
```
Comment (pedrobaeza) [inline addons/account_voucher/migrations/9.0.1.0/post-migration.py]: Do you really need the subquery?

Comment (AaronHForgeFlow) [reply addons/account_voucher/migrations/9.0.1.0/post-migration.py]: I couldn't find a way

File: addons/account_voucher/migrations/9.0.1.0/post-migration.py [hunk 3]
```diff
@@ -83,15 +83,54 @@ def create_payments_from_vouchers(env):
         AND aml.id = aml2.id
         """
     )
+
+    if not openupgrade.column_exists(
+            env.cr, 'account_move_line', 'payment_from_invoice_id'):
```
Comment (StefanRijnhart) [inline addons/account_voucher/migrations/9.0.1.0/post-migration.py]: Can you use `openupgrade.get_legacy_name('payment_from_invoice_id')` so that the legacy character of this table is recognizable at database cleanup time?

File: addons/account_voucher/migrations/9.0.1.0/post-migration.py [hunk 4]
```diff
@@ -83,11 +83,50 @@ def create_payments_from_vouchers(env):
...
+    env.cr.execute("""SELECT column_name
+        FROM information_schema.columns
+        WHERE table_name='account_move_line' AND
+        column_name='payment_from_invoice_id'""")
+    if not env.cr.fetchone():
+        env.cr.execute("ALTER TABLE account_move_line ADD COLUMN"
+                       "payment_from_invoice_id INTEGER")
+
+    env.cr.execute(
+        """\
```
Comment (mreficent) [inline addons/account_voucher/migrations/9.0.1.0/post-migration.py]: delete the `\`

File: addons/account_voucher/migrations/9.0.1.0/post-migration.py [hunk 5]
```diff
@@ -83,17 +83,57 @@ def create_payments_from_vouchers(env):
...
+            env.cr, 'account_move_line',
+            openupgrade.get_legacy_name('payment_from_invoice_id')):
+        new = openupgrade.get_legacy_name('payment_from_invoice_id')
+        env.cr.execute("""
+            ALTER TABLE account_move_line ADD COLUMN
+            %s INTEGER
+            """ % new)
+
     env.cr.execute(
         """\
```
Comment (MiquelRForgeFlow) [inline addons/account_voucher/migrations/9.0.1.0/post-migration.py]: delete the `\`

File: addons/account_voucher/migrations/9.0.1.0/post-migration.py [hunk 6]
```diff
@@ -83,17 +83,57 @@ def create_payments_from_vouchers(env):
...
+                    SELECT aml_id FROM Q1 WHERE av_id = av.id
+                )
+            )
+        )""" % openupgrade.get_legacy_name('payment_from_invoice_id'))
+
     # Also recreate link from invoice to payment
     env.cr.execute(
-        """\
+        """
         INSERT INTO account_invoice_payment_rel (payment_id, invoice_id)
```
Comment (StefanRijnhart) [inline addons/account_voucher/migrations/9.0.1.0/post-migration.py:89]: It seems to me that you can fill this table directly from the following query:
```
SELECT av.id as payment_id, ai.id as invoice_id
FROM account_invoice ai
JOIN account_move_line amli ON ai.move_id = amli.move_id
JOIN account_partial_reconcile apr ON 
     (apr.credit_move_id = amli.id OR apr.debit_move_id = amli.id)
JOIN account_move_line amlp ON
     (apr.credit_move_id = amlp.id OR apr.debit_move_id = amlp.id)
JOIN account_voucher av ON av.move_id = amlp.move_id
WHERE av.voucher_type IN ('receipt', 'payment')
AND av.state IN ('draft', 'posted');
```
I might be missing something though, so please have a careful look. If it works, it would be good because it is a bit quicker than the query above.

Comment (AaronHForgeFlow) [reply addons/account_voucher/migrations/9.0.1.0/post-migration.py:89]: Thanks. I'll test this one and I'll come back with the result

Comment (AaronHForgeFlow) [reply addons/account_voucher/migrations/9.0.1.0/post-migration.py:89]: Tested with slight modification. Same result as previous approach. Thank you!

File: addons/account_voucher/migrations/9.0.1.0/post-migration.py [hunk 7]
```diff
@@ -85,13 +85,19 @@ def create_payments_from_vouchers(env):
...
-        GROUP BY aml.payment_id, ai.id
+        SELECT av.id as payment_id, ai.id as invoice_id
+        FROM account_invoice ai
+        JOIN account_move_line amli ON ai.move_id = amli.move_id
+        JOIN account_partial_reconcile apr ON 
+             (apr.credit_move_id = amli.id OR apr.debit_move_id = amli.id)
+        JOIN account_move_line amlp ON
+             (apr.credit_move_id = amlp.id OR apr.debit_move_id = amlp.id)
+        JOIN account_voucher av ON av.move_id = amlp.move_id
+        WHERE av.voucher_type IN ('sale', 'purchase')
```
Comment (StefanRijnhart) [inline addons/account_voucher/migrations/9.0.1.0/post-migration.py]: Why did you change this to sale or purchase type vouchers? Only vouchers with 'receipt' or 'payment' are migrated to account.payment in the first place. type https://github.com/OCA/OpenUpgrade/blob/9.0/addons/account_voucher/migrations/9.0.1.0/post-migration.py#L55

Comment (AaronHForgeFlow) [reply addons/account_voucher/migrations/9.0.1.0/post-migration.py]: I'm still reviewing. It came out at the time the post-migration script was executed that the voucher_type was already updated to the selection values for v9. In that case the condition is useless because it's fetching all vouchers of all types. What I don't know is why the voucher_type column was updated when there is no script for that.

Comment (StefanRijnhart) [reply addons/account_voucher/migrations/9.0.1.0/post-migration.py]: Any news on this?

**Comment by pedrobaeza** [comment]
@aheficent do you finish this?

**Comment by AaronHForgeFlow** [comment]
@pedrobaeza @StefanRijnhart Yes I will continue with this, sorry for the delay. I've using this script for a custom migration and it seems to be working but I'll double check to make sure it works in all cases.

PS: If anybody has used this Pull Request for testing a migration it'll be great to get some feedback here. :grimacing:

**Comment by AaronHForgeFlow** [comment]
@StefanRijnhart you are right. The previous solution was not working. Besides, it makes sense to use the same voucher_type as the one used to create the payments.

To sum up, this is working fine. Tested migration in a database with more than 10000 payments. Selected   random invoices and the payment was successfully linked.

Sorry for the delay.

**Review by StefanRijnhart — APPROVED** [review]

**Comment by StefanRijnhart** [comment]
There has been a visit from the trailing whitespace monster.

**Comment by AaronHForgeFlow** [comment]
@pedrobaeza @mreficent
